crates/turbo-tasks/src/invalidation_reason_set.rs

```diff
@@ -0,0 +1,113 @@
+use std::{borrow::Cow, fmt::Display, mem::take};
+
+use indexmap::{map::Entry, IndexMap, IndexSet};
+
+use crate::{
+    manager::{InvalidationReason, InvalidationReasonType},
+    util::StaticOrArc,
+};
+
+#[derive(PartialEq, Eq, Hash)]
+enum MapKey {
+    Untyped { unique_tag: usize },
+    Typed { type_pointer: usize },
+}
+
+enum MapEntry {
+    Untyped {
+        reason: StaticOrArc<dyn InvalidationReason>,
+    },
+    Single {
+        reason: StaticOrArc<dyn InvalidationReason>,
+        data: Cow<'static, str>,
+    },
+    Multiple {
+        reason_ty: &'static dyn InvalidationReasonType,
+        merge_data: IndexSet<Cow<'static, str>>,
+    },
+}
+
+#[derive(Default)]
+pub struct InvalidationReasonSet {
+    next_unique_tag: usize,
+    map: IndexMap<MapKey, MapEntry>,
```

Comment:
There should be two maps: a typed one, and an untyped one. Otherwise types get muddy below and you need an `unreachable`.

```suggestion
    typed_map: IndexMap<MapKey, MapEntry>,
    untyped_map: IndexMap<MapKey, MapEntry>,
```

Reply:
I initially had that, but that won't keep the order of reasons intact when mixing typed and untyped reasons.

Reply:
I see. You could add an ordering `idx` to each entry but that might have its own issues. Let's add a comment to explain why everything is in a single map.